### PR TITLE
Added eps attribute to FrozenBatchNorm2d

### PR DESCRIPTION
**Pull request by frgfm**
This PR aims at tackling #2169 by:
- adding a `eps` attribute to  ̀FrozenBatchNorm2d`
- renaming the `n` positional argument of  `FrozenBatchNorm2d` constructor to `num_features`
- adding a unittest to ensure default eps is zero for backward compatibility
- adding a unittest to ensure valid computation when epsilon is higher than zero by comparison with `BatchNorm2d` forward in eval mode.

### CODE REVIEW

**Comment by frgfm** [comment]
@fmassa I'm not sure where the error comes from, it seems like the tests timed out in `test_ops` at `NMSTester`. Do you think it could be due to this PR?

**Comment by frgfm** [comment]
Also, I took two choices here that any reviewer might want to take a look at:
- renaming the constructor argument `n` to get closer to `BatchNorm2d` interface (I didn't see many reasons for someone to use kwargs, but it might be worth mentioning it as a breaking changes if merged into the next release)

- rearranging the `forward` implementation by switching `reshape` to `view` (I might have missed a use case where it's needed), and delaying this reshape/view to the last moment to reduce the LoC. I made tests to ensure the computation is the exact same as previously, and I also checked execution time on both CPU & GPU which were the same.

File: test/test_ops.py [hunk 1]
```diff
@@ -547,6 +547,30 @@ def test_frozenbatchnorm2d_repr(self):
...
+        bn.load_state_dict(state_dict)
+        # Difference is expected to fall in an acceptable range
+        self.assertTrue(torch.allclose(fbn(x), bn(x), atol=1e-6))
+
+        # Check computation for eps > 0
+        fbn = ops.misc.FrozenBatchNorm2d(sample_size[1], eps=1e-5)
+        fbn.load_state_dict(state_dict, strict=False)
+        bn = torch.nn.BatchNorm2d(sample_size[1], eps=1e-5).eval()
+        bn.load_state_dict(state_dict)
+        self.assertTrue(torch.allclose(fbn(x), bn(x), atol=1e-6))
```
Comment (zhangguanheng66) [inline test/test_ops.py:572]: Should be use `torch.testing.assert_allclose` here?

**Comment by zhangguanheng66** [comment]
> Also, I made two choices here that any reviewer might want to take a look at:
> 
> * renaming the constructor argument `n` to get closer to `BatchNorm2d` interface (I didn't see many reasons for someone to use kwargs, but it might be worth mentioning it as a breaking changes if merged into the next release)
> * rearranging the `forward` implementation by switching `reshape` to `view` (I might have missed a use case where it's needed), and delaying this reshape/view to the last moment to reduce the LoC. I made tests to ensure the computation is the exact same as previously, and I also checked execution time on both CPU & GPU which were the same.

For the first option, I'm not sure if it introduces a BC breaking because it just re-defines the name of the argument.

File: test/test_ops.py (hunk 1 above)
Comment (frgfm) [reply test/test_ops.py:572]: Oh I didn't know about it!
So you're suggesting `self.assertIsNone(torch.testing.assert_allclose(fbn(x), bn(x), atol=1e-6))` ? Or you think we should drop the unittest interface?

Comment (fmassa) [reply test/test_ops.py:572]: The current implementation with `assertTrue` is fine. At one point in the future we might want to take the custom Testing class from PyTorch (which defines equality between tensors within `assertEqual`), but I kind of prefer using the tester methods instead of raising asserts in tests.

File: torchvision/ops/misc.py [hunk 2]
```diff
@@ -142,15 +143,12 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
...
     def forward(self, x):
-        # move reshapes to the beginning
-        # to make it fuser-friendly
-        w = self.weight.reshape(1, -1, 1, 1)
-        b = self.bias.reshape(1, -1, 1, 1)
-        rv = self.running_var.reshape(1, -1, 1, 1)
-        rm = self.running_mean.reshape(1, -1, 1, 1)
-        scale = w * rv.rsqrt()
-        bias = b - rm * scale
-        return x * scale + bias
```
Comment (fmassa) [inline torchvision/ops/misc.py:153]: Can you revert those changes? It was necessary to get much faster runtime when running with the JIT fuser (it was all done in a single kernel call), and I'm not sure the fuser is already smart enough to re-order those operations so that the we don't need to do it ourselves.

cc @suo for an example where we had to move operations to the beginning to facilitate fusion. Note that in this case one complicating factor is the reshapes and the broadcasting

File: torchvision/ops/misc.py [hunk 3]
```diff
@@ -142,15 +143,12 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
...
+        # Scaling factor
+        scale = self.weight * (self.running_var + self.eps).rsqrt()
+        # Bias
+        bias = (self.bias - self.running_mean * scale)
+
+        return x * scale.view(1, -1, 1, 1) + bias.view(1, -1, 1, 1)
```
Comment (fmassa) [inline torchvision/ops/misc.py:151]: In general, I prefer to use `reshape` instead of `view`, because it's safer: `view` can raise errors if the tensor is not contiguous, while `reshape` makes it contiguous beforehand. Plus, `reshape` is smart so that it doesn't make unnecessary copies, so for me it is strictly better than `view`.

File: torchvision/ops/misc.py [hunk 4]
```diff
@@ -124,12 +124,13 @@ class FrozenBatchNorm2d(torch.nn.Module):
     are fixed
     """
 
-    def __init__(self, n):
+    def __init__(self, num_features, eps=0.):
```
Comment (fmassa) [inline torchvision/ops/misc.py:127]: for backwards-compatibility, can you also define a `n=None` kwarg?
Something like
```python
def __init__(self, num_features, eps=0.0, n=None):
    if n is not None:
        warnings.warn("n is deprecated and has been renamed num_features")
        num_features = n
```

File: torchvision/ops/misc.py (hunk 2 above)
Comment (frgfm) [reply torchvision/ops/misc.py]: Yes I wasn't sure about JIT, thanks for the details. Just reverted the changes!

File: torchvision/ops/misc.py (hunk 4 above)
Comment (frgfm) [reply torchvision/ops/misc.py]: Just added it, thanks for the suggestion!

**Comment by fmassa** [comment]
Test failures seem related

**Comment by frgfm** [comment]
Thanks, I fixed those. There is a timeout remaining for windows builds and an unrelated (I think) failure on travis!
